File: packages/react-native-hms/ios/HMSRNInteractivityCenter.swift
//
//  HMSRNInteractivityCenter.swift
//  react-native-hms
//
//  Created by Jatin Nagar on 09/01/24.
//

import Foundation
import HMSSDK

class HMSRNInteractivityCenter {
    weak var hmssdk: HMSSDK?
    weak var hmsrnsdk: HMSRNSDK?

    init(hmssdk: HMSSDK, hmsrnsdk: HMSRNSDK?) {
        self.hmssdk = hmssdk
        self.hmsrnsdk = hmsrnsdk

        hmssdk.interactivityCenter.addPollUpdateListner { [weak self] updatedPoll, update in
            guard let self = self else { return }
            guard let enabledEvents = self.hmsrnsdk?.eventsEnableStatus, enabledEvents[HMSConstants.ON_POLL_UPDATE] == true else {
                print("HMSConstants.ON_POLL_UPDATE event is not enabled")
                return
            }
            self.hmsrnsdk?.delegate?.emitEvent(
                HMSConstants.ON_POLL_UPDATE,
                ["update": update.rawValue, "updatedPoll": HMSInteractivityDecoder.getHMSPoll(updatedPoll)]
            )
      }
    }

    func quickStartPoll(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let pollBuilder = HMSInteractivityHelper.getPollBuilderFromDict(data, sdkRoles: hmssdk?.roles)

        self.hmssdk?.interactivityCenter.quickStartPoll(with: pollBuilder) { _, error in
            if let nonnilError = error {
                reject?("6004", nonnilError.localizedDescription, nil)
                return
            }
            resolve?(nil)
        }
    }

    func add(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        guard let pollId = data["pollId"] as? String,
              let poll = self.hmssdk?.interactivityCenter.polls.first(where: {poll in poll.pollID == pollId}) else {
            reject?("6004", "Unable to find HMSPoll with given pollId", nil)
            return
        }
        guard let pollQuestionIndex = data["pollQuestionIndex"] as? Int,
              let pollQuestion = poll.questions?.first(where: {question in question.index == pollQuestionIndex}) else {
            reject?("6004", "Unable to find HMSPollQuestion in poll with given question index", nil)
            return
        }
        guard let responses = data["responses"] as? NSDictionary else {
            reject?("6004", "responses field is required", nil)
            return
        }

        let pollResponseBuilder = HMSInteractivityHelper.getPollResponseBuilder(responses, poll: poll, pollQuestion: pollQuestion)

        self.hmssdk?.interactivityCenter.add(response: pollResponseBuilder) { pollQuestionResponseResult, error in
            if let nonnilError = error {
                reject?("6004", nonnilError.localizedDescription, nil)
                return
            }
            if let pollQuestionResponseResult = pollQuestionResponseResult {
                resolve?(HMSInteractivityDecoder.getHMSPollQuestionResponseResults(pollQuestionResponseResult))
            } else {
                resolve?(nil)
            }
        }
    }

    func stop(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        guard let pollId = data["pollId"] as? String,
              let poll = self.hmssdk?.interactivityCenter.polls.first(where: {poll in poll.pollID == pollId}) else {
            reject?("6004", "Unable to find HMSPoll with given pollId", nil)
            return
        }
        self.hmssdk?.interactivityCenter.stop(poll: poll, completion: {success, error in
            if let nonnilError = error {
                reject?("6004", nonnilError.localizedDescription, nil)
                return
            }
            resolve?(success)
        })
    }

    func fetchLeaderboard(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        guard let pollId = data["pollId"] as? String,
              let poll = self.hmssdk?.interactivityCenter.polls.first(where: {poll in poll.pollID == pollId}) else {
            reject?("6004", "Unable to find HMSPoll with given pollId", nil)
            return
        }

        guard let count = data["count"] as? Int,
            let startIndex = data["startIndex"] as? Int
        else {
            reject?("6004", "Unable to find required parameters", nil)
            return
        }

        let includeCurrentPeer = data["includeCurrentPeer"] as? Bool ?? false

        self.hmssdk?.interactivityCenter.fetchLeaderboard(for: poll, offset: startIndex, count: count, includeCurrentPeer: includeCurrentPeer) { response, error in

            if let nonnilError = error {
                reject?("6004", nonnilError.localizedDescription, nil)
                return
            }

            if let response = response {
                resolve?(HMSInteractivityDecoder.getLeaderboardResponse(response))
            } else {
                reject?("6004", "Could not fetch leaderboard response", nil)
            }
        }
    }
}

File: packages/react-native-hms/android/src/main/java/com/reactnativehmssdk/Interactivity/HMSRNInteractivityCenter.kt
package com.reactnativehmssdk
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableMap
import com.reactnativehmssdk.Interactivity.HMSInteractivityDecoder
import live.hms.video.error.HMSException
import live.hms.video.interactivity.HmsPollUpdateListener
import live.hms.video.polls.models.HMSPollUpdateType
import live.hms.video.polls.models.HmsPoll
import live.hms.video.polls.models.answer.PollAnswerResponse
import live.hms.video.polls.network.PollLeaderboardResponse
import live.hms.video.sdk.HMSActionResultListener
import live.hms.video.sdk.HMSSDK
import live.hms.video.sdk.HmsTypedActionResultListener
import live.hms.video.whiteboard.HMSWhiteboardUpdate
import live.hms.video.whiteboard.HMSWhiteboardUpdateListener

class HMSRNInteractivityCenter(private val sdk: HMSSDK, private val rnSDK: HMSRNSDK) {
  init {
    //region Listen for poll updates
    this.sdk.getHmsInteractivityCenter().pollUpdateListener =
      object : HmsPollUpdateListener {
        override fun onPollUpdate(
          hmsPoll: HmsPoll,
          hmsPollUpdateType: HMSPollUpdateType,
        ) {
          if (rnSDK.eventsEnableStatus["ON_POLL_UPDATE"] != true) {
            return
          }

          val data: WritableMap = Arguments.createMap()

          data.putInt("update", HMSInteractivityDecoder.getPollUpdateType(hmsPollUpdateType))
          data.putMap("updatedPoll", HMSInteractivityDecoder.getPoll(hmsPoll))

          rnSDK.delegate.emitEvent("ON_POLL_UPDATE", data)
        }
      }
    //endregion

    //region Listen for whiteboard updates
    this.sdk.getHmsInteractivityCenter().setWhiteboardUpdateListener(
      object : HMSWhiteboardUpdateListener {
        override fun onUpdate(hmsWhiteboardUpdate: HMSWhiteboardUpdate) {
          if (rnSDK.eventsEnableStatus["ON_WHITEBOARD_UPDATE"] != true) {
            return
          }
          val data: WritableMap = Arguments.createMap()
          when (hmsWhiteboardUpdate) {
            is HMSWhiteboardUpdate.Start -> {
              data.putMap("hmsWhiteboard", HMSInteractivityDecoder.getHMSWhiteboard(hmsWhiteboardUpdate.hmsWhiteboard))
              data.putString("updateType", HMSInteractivityDecoder.getWhiteboardUpdateType(hmsWhiteboardUpdate))
            }
            is HMSWhiteboardUpdate.Stop -> {
              data.putMap("hmsWhiteboard", HMSInteractivityDecoder.getHMSWhiteboard(hmsWhiteboardUpdate.hmsWhiteboard))
              data.putString("updateType", HMSInteractivityDecoder.getWhiteboardUpdateType(hmsWhiteboardUpdate))
            }
          }
          rnSDK.delegate.emitEvent("ON_WHITEBOARD_UPDATE", data)
        }
      },
    )
    //endregion
  }

  //region poll methods
  fun quickStartPoll(
    data: ReadableMap,
    promise: Promise?,
  ) {
    this.sdk.getHmsInteractivityCenter().quickStartPoll(
      HMSInteractivityHelper.getPollBuilder(data, sdk.getRoles()),
      object : HMSActionResultListener {
        override fun onError(error: HMSException) {
          // Error
          promise?.reject(error.message, error)
        }

        override fun onSuccess() {
          // Success
          promise?.resolve(null)
        }
      },
    )
  }

  fun addResponseOnPollQuestion(
    data: ReadableMap,
    promise: Promise?,
  ) {
    val pollId = data.getString("pollId")
    if (pollId == null) {
      promise?.reject("6004", "pollId is required")
      return
    }
    val poll = this.sdk.getHmsInteractivityCenter().polls.find { it.pollId == pollId }
    if (poll == null) {
      promise?.reject("6004", "No HMSPoll with pollId `$pollId`")
      return
    }
    val pollQuestionIndex = data.getInt("pollQuestionIndex")
    val pollQuestion = poll.questions?.find { it.questionID == pollQuestionIndex }
    if (pollQuestion == null) {
      promise?.reject("6004", "No HMSPollQuestion in poll with given question index")
      return
    }
    val responses = data.getMap("responses")
    if (responses == null) {
      promise?.reject("6004", "responses field is required")
      return
    }
    val pollResponseBuilder = HMSInteractivityHelper.getPollResponseBuilder(responses, poll, pollQuestion)
    this.sdk.getHmsInteractivityCenter().add(
      pollResponseBuilder,
      object : HmsTypedActionResultListener<PollAnswerResponse> {
        override fun onError(error: HMSException) {
          promise?.reject(error.code.toString(), error.description)
        }

        override fun onSuccess(result: PollAnswerResponse) {
          promise?.resolve(HMSInteractivityDecoder.getHMSPollQuestionResponseResults(result))
        }
      },
    )
  }

  fun stop(
    data: ReadableMap,
    promise: Promise?,
  ) {
    val pollId = data.getString("pollId")
    if (pollId == null) {
      promise?.reject("6004", "pollId is required")
      return
    }
    val poll = this.sdk.getHmsInteractivityCenter().polls.find { it.pollId == pollId }
    if (poll == null) {
      promise?.reject("6004", "No HMSPoll with pollId `$pollId`")
      return
    }
    this.sdk.getHmsInteractivityCenter().stop(
      poll,
      object : HMSActionResultListener {
        override fun onError(error: HMSException) {
          promise?.reject(error.code.toString(), error.description)
        }

        override fun onSuccess() {
          promise?.resolve(true)
        }
      },
    )
  }

  fun fetchLeaderboard(
    data: ReadableMap,
    promise: Promise?,
  ) {
    val pollId = data.getString("pollId")
    if (pollId == null) {
      promise?.reject("6004", "pollId is required")
      return
    }
    val poll = this.sdk.getHmsInteractivityCenter().polls.find { it.pollId == pollId }
    if (poll == null) {
      promise?.reject("6004", "No HMSPoll with pollId `$pollId`")
      return
    }
    val count = data.getInt("count")
    val startIndex = data.getInt("startIndex")
    val includeCurrentPeer = data.getBoolean("includeCurrentPeer")

    this.sdk.getHmsInteractivityCenter().fetchLeaderboard(
      pollId = poll.pollId,
      count = count.toLong(),
      startIndex = startIndex.toLong(),
      includeCurrentPeer = includeCurrentPeer,
      object : HmsTypedActionResultListener<PollLeaderboardResponse> {
        override fun onSuccess(result: PollLeaderboardResponse) {
          promise?.resolve(HMSInteractivityDecoder.getPollLeaderboardResponse(result))
        }

        override fun onError(error: HMSException) {
          promise?.reject(error.code.toString(), error.description)
        }
      },
    )
  }
  //endregion

  //region whiteboard methods
  fun startWhiteboard(
    data: ReadableMap,
    promise: Promise?,
  ) {
    val whiteboardTitle = data.getString("title")
    if (whiteboardTitle == null) {
      promise?.reject("6004", "whiteboard title is required")
      return
    }
    sdk.getHmsInteractivityCenter().startWhiteboard(
      whiteboardTitle,
      object : HMSActionResultListener {
        override fun onSuccess() {
          promise?.resolve(true)
        }

        override fun onError(error: HMSException) {
          promise?.reject(error.code.toString(), error.description)
        }
      },
    )
  }

  fun stopWhiteboard(promise: Promise?) {
    sdk.getHmsInteractivityCenter().stopWhiteboard(
      object : HMSActionResultListener {
        override fun onSuccess() {
          promise?.resolve(true)
        }

        override fun onError(error: HMSException) {
          promise?.reject(error.code.toString(), error.description)
        }
      },
    )
  }
  //endregion
}
